src/main/scala/mesosphere/marathon/core/health/impl/HealthIndex.scala

```diff
@@ -0,0 +1,99 @@
+package mesosphere.marathon
+package core.health.impl
+
+import mesosphere.marathon.core.health.Health
+import mesosphere.marathon.core.instance.Instance
+import mesosphere.marathon.core.task.Task
+
+import scala.collection.concurrent.TrieMap
+
+trait HealthIndex {
+  def remove(key: Instance.Id): Unit
+
+  def remove(key: Task.Id): Unit
+
+  def +=(kv: (Task.Id, Health)): this.type
+
+  def values: Iterable[Health]
+
+  def filterKeys(p: Task.Id => Boolean): collection.Map[Task.Id, Health]
+
+  def get(key: Instance.Id): Option[Health]
+
+  def getOrElse(key: Task.Id, default: => Health): Health
+
+  def retain(p: Task.Id => Boolean): this.type
+}
+
+/** Old implementation, search by Instance.Id is slow */
+case class HealthSingleTrieMap() extends HealthIndex {
+  val healthByTaskId: TrieMap[Task.Id, Health] = TrieMap.empty[Task.Id, Health]
+
+  def remove(key: Instance.Id): Unit = get(key).foreach(h => remove(h.instanceId))
+
+  def remove(key: Task.Id): Unit = healthByTaskId.remove(key)
+
+  /** Slow, full scan algorithm O(n) */
+  def get(key: Instance.Id): Option[Health] = healthByTaskId.find(_._1.instanceId == key).map(_._2)
+
+  def getOrElse(key: Task.Id, default: => Health): Health = healthByTaskId.getOrElse(key, default)
+
+  def filterKeys(p: Task.Id => Boolean): collection.Map[Task.Id, Health] = {
+    healthByTaskId.filterKeys(p)
+  }
+
+  /** Slow */
+  def retain(p: Task.Id => Boolean): this.type = {
+    healthByTaskId.retain((taskId, _) =>  p(taskId))
+    this
+  }
+
+  def +=(kv: (Task.Id, Health)): this.type = {
+    healthByTaskId += kv
+    this
+  }
+
+  def values: Iterable[Health] = healthByTaskId.values
+}
+
+/** New implementation, search by Instance.Id is fast but we need two indexes */
+case class HealthDualTrieMap() extends HealthIndex {
+  val healthByTaskId: TrieMap[Task.Id, Health] = TrieMap.empty[Task.Id, Health]
+  val healthByInstanceId: TrieMap[Instance.Id, Health] = TrieMap.empty[Instance.Id, Health]
+
+  def remove(key: Instance.Id): Unit = get(key).foreach(h => remove(h.instanceId))
```

Comment:
Maybe I'm missing something obvious or a particular scala think, but I don't understand this method; especially, how it works to call remove on an instance Id, whereis it seems to be this method itself 🤔 should not we call remove(Task.Id) instead? So that both Trie are emptied from the entry ?

Reply:
I think that's the purpose of the next method, complete remove of a task. If your remove a task, you can clean up everything, but if you remove an instance id, you **might** still need the task info (unless it's the last known instance id linked to this task?).
Anyway, why are we discussing this? The method is not used apparently and likely to create consistency bugs, shouldn't we just drop it??

Reply:
I still don't understand how the first one is supposed to work, as it calls remove on an Instande.Id, so this exact same method again and again recursively; no ?

Reply:
It was a bug, fixed now.